t/chaos/delayetcd/delayetcd.go

```diff
@@ -0,0 +1,191 @@
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/apache/apisix/t/chaos/utils"
+)
+
+func getEtcdDelayChaos(delay int) *v1alpha1.NetworkChaos {
+	return &v1alpha1.NetworkChaos{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "etcd-delay",
+			Namespace: metav1.NamespaceDefault,
+		},
+		Spec: v1alpha1.NetworkChaosSpec{
+			Selector: v1alpha1.SelectorSpec{
+				LabelSelectors: map[string]string{"app.kubernetes.io/instance": "etcd"},
+			},
+			Action: v1alpha1.DelayAction,
+			Mode:   v1alpha1.AllPodMode,
+			TcParameter: v1alpha1.TcParameter{
+				Delay: &v1alpha1.DelaySpec{
+					Latency: strconv.Itoa(delay) + "ms",
+				},
+			},
+		},
+	}
+}
+
+func setRouteMultipleTimes(e *httpexpect.Expect, times int, status httpexpect.StatusRange) time.Duration {
+	now := time.Now()
+	timeLast := now
+	var timeList []string
+	for i := 0; i < times; i++ {
+		utils.SetRoute(e, status)
+		timeList = append(timeList, time.Since(timeLast).String())
+		timeLast = time.Now()
+	}
+	fmt.Fprintf(ginkgo.GinkgoWriter, "takes %v separately", timeList)
+	return time.Since(now) / time.Duration(times)
+}
+
+var _ = ginkgo.Describe("Test APISIX Delay When Add ETCD Delay", func() {
+	ctx := context.Background()
+	e := httpexpect.New(ginkgo.GinkgoT(), utils.Host)
+	eSilent := utils.GetSilentHttpexpectClient()
+
+	var cliSet *utils.ClientSet
+	var apisixPod *v1.Pod
+	var err error
+	ginkgo.It("init client set", func() {
+		cliSet, err = utils.InitClientSet()
+		gomega.Expect(err).To(gomega.BeNil())
+		listOption := client.MatchingLabels{"app": "apisix-gw"}
+		apisixPods, err := utils.GetPods(cliSet.CtrlCli, metav1.NamespaceDefault, listOption)
+		gomega.Expect(err).To(gomega.BeNil())
+		gomega.Ω(len(apisixPods)).Should(gomega.BeNumerically(">", 0))
+		apisixPod = &apisixPods[0]
+	})
+
+	defer ginkgo.It("restore test environment", func() {
+		utils.DeleteRoute(e)
+	})
+
+	ginkgo.It("check if everything works", func() {
+		utils.SetRoute(e, httpexpect.Status2xx)
+		utils.GetRouteList(e, http.StatusOK)
+
+		utils.CheckMethodSucceed(e, http.MethodGet, 1)
+		utils.TestPrometheusEtcdMetric(e, 1)
+	})
+
+	// get default
+	ginkgo.It("get default apisix delay", func() {
+		timeStart := time.Now()
+		setDuration := setRouteMultipleTimes(eSilent, 5, httpexpect.Status2xx)
+		gomega.Ω(setDuration).Should(gomega.BeNumerically("<", 15*time.Millisecond))
+
+		errorLog, err := utils.Log(apisixPod, cliSet.KubeCli, timeStart)
+		gomega.Expect(err).To(gomega.BeNil())
+		gomega.Ω(errorLog).ShouldNot(gomega.ContainSubstring("error"))
+	})
+
+	// 30ms delay
+	ginkgo.It("generate a 30ms delay between etcd and apisix", func() {
+		timeStart := time.Now()
+		chaos := getEtcdDelayChaos(30)
+		err := cliSet.CtrlCli.Create(ctx, chaos)
+		gomega.Expect(err).To(gomega.BeNil())
+		time.Sleep(1 * time.Second)
+
+		defer func() {
+			err = cliSet.CtrlCli.Delete(ctx, chaos)
+			gomega.Expect(err).To(gomega.BeNil())
+			time.Sleep(1 * time.Second)
```

Comment:
The sleep mechanism might not be so reliable, I'm afraid it's unstable in the Github Action.

We may try to make sure the environment is clean before each case runs.

Reply:
Thanks fixed. PTAL